packages/flutter/lib/src/widgets/text_editing_action.dart

```diff
@@ -19,12 +22,847 @@ import 'framework.dart';
+    if (nextSelection == value.selection) {
+      return;
+    }
+    final int distance = nextSelection.extentOffset - value.selection.extentOffset;
+    _cursorResetLocation += distance;
+    setSelection(nextSelection, cause);
+  }
+
+  /// {@macro flutter.rendering.TextEditingValue.extendSelectionRightByLine}
+  ///
+  /// If [selectionEnabled] is false, keeps the selection collapsed and moves it
+  /// right by line.
+  ///
+  /// {@macro flutter.rendering.RenderEditable.cause}
+  ///
+  /// See also:
+  ///
+  ///   * [TextEditingValue.extendSelectionRightByLine], which is used by this method.
+  ///   * [extendSelectionLeftByLine], which is same but in the opposite
+  ///     direction.
+  ///   * [expandSelectionRightByLine], which strictly grows the selection
+  ///     regardless of the order.
+  void extendSelectionRightByLine(SelectionChangedCause cause) {
+    if (!renderEditable.selectionEnabled) {
+      return moveSelectionRightByLine(cause);
+    }
+
+    setSelection(value.extendSelectionRightByLine(renderEditable), cause);
+  }
+
+  // Extend the current selection to the start of the field.
+  //
+  // If selectionEnabled is false, keeps the selection collapsed and moves it to
+  // the start.
+  //
+  // The given [SelectionChangedCause] indicates the cause of this change and
+  // will be passed to [onSelectionChanged].
+  //
+  // See also:
+  //
+  //   * _extendSelectionToEnd
+  void _extendSelectionToStart(SelectionChangedCause cause) {
+    if (!renderEditable.selectionEnabled) {
+      return moveSelectionToStart(cause);
+    }
+
+    final TextSelection nextSelection = value.extendSelectionToStart();
+    setSelection(nextSelection, cause);
+  }
+
+  /// {@macro flutter.rendering.TextEditingValue.extendSelectionLeftByWord}
+  ///
+  /// Extend the current [selection] to the previous start of a word.
+  ///
+  /// {@macro flutter.rendering.RenderEditable.cause}
+  ///
+  /// {@macro flutter.rendering.RenderEditable.whiteSpace}
+  ///
+  /// {@template flutter.rendering.RenderEditable.stopAtReversal}
+  /// The `stopAtReversal` parameter is false by default, meaning that it's
+  /// ok for the base and extent to flip their order here. If set to true, then
+  /// the selection will collapse when it would otherwise reverse its order. A
+  /// selection that is already collapsed is not affected by this parameter.
+  /// {@endtemplate}
+  ///
+  /// See also:
+  ///
+  ///   * [TextEditingValue.extendSelectionLeftByWord], which is used by this method.
+  ///   * [extendSelectionRightByWord], which is the same but in the opposite
+  ///     direction.
+  void extendSelectionLeftByWord(SelectionChangedCause cause, [bool includeWhitespace = true, bool stopAtReversal = false]) {
+    // When the text is obscured, the whole thing is treated as one big word.
+    if (obscureText) {
+      return _extendSelectionToStart(cause);
+    }
+
+    // TODO(justinmc): I think this assert has to happen in RenderEditable. If
```

Comment:
This is probably one of the big things I still need to do.  I think that doing the assert in RenderEditable as I described in this comment is doable.  I kind of wish we didn't need this assertion at all, though.